File: javascript/bob/bob.js
'use strict';

function Bob () {}

Bob.prototype.hey = function (message) {
    var response = 'Whatever.',
        has_letter = /[a-zA-Z]/.test(message),
        is_empty = !message.replace(/\s+/, ''),
        is_question = message.substr(-1) === '?',
        is_upper = message === message.toUpperCase();

    if (is_empty) {
        response = 'Fine. Be that way!';
    } else if (has_letter && is_upper) {
        response = 'Woah, chill out!';
    } else if (is_question) {
        response = 'Sure.';
    }

    return response;
};

module.exports = Bob;

File: Exercism/scala/bob/src/main/scala/Bob.scala
class Bob {
  def isQuestion(msg: String): Boolean =
    msg.endsWith("?")

  def isNothing(msg: String): Boolean =
    msg.trim.isEmpty

  def isYelling(msg: String): Boolean = {
    val letters = msg.filter(_.isLetter)
    letters.forall(_.isUpper) && !letters.isEmpty
  }

  def hey(msg: String): String =
    if (isNothing(msg))
      "Fine. Be that way!"
    else if (isYelling(msg))
      "Whoa, chill out!"
    else if (isQuestion(msg))
      "Sure."
    else
      "Whatever."
}
